internal/services/rate.go

```diff
@@ -1,115 +1,160 @@
+	ctx, cancel := context.WithTimeout(context.Background(), RateFetchTimeout)
+	defer cancel()
+
+	reqURL := fmt.Sprintf("%s/%s.json", FawazAhmedExchangeRateURL, strings.ToLower(code))
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
+	if err != nil {
+		return 0, err
+	}
 
-type ExchangeRateClient interface {
-	GetExchangeRate() (float32, error)
+	resp, err := client.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		return 0, ErrInvalidCurrencyCode
+	}
+
+	defer resp.Body.Close()
+	var fawazAhmedResponse FawazAhmedResponse
+	err = json.NewDecoder(resp.Body).Decode(&fawazAhmedResponse)
+	if err != nil {
+		return 0, err
+	}
+	return fawazAhmedResponse.Rates.UAH, nil
 }
 
 type cachingRateService struct {
-	currentRate    float32
-	mutex          sync.RWMutex
-	fetchError     error
+	namedFetchers  []NamedFetcher
+	client         *http.Client
 	updateInterval time.Duration
-	client         ExchangeRateClient
+	cache          *cache.Cache[string, float32]
 }
 
-// Creates new rate service instance.
-// Pass nil to client to use default http client.
-//
-//nolint:revive
-func NewRateService(updateInterval time.Duration, client ExchangeRateClient) *cachingRateService {
-	return &cachingRateService{
-		mutex:          sync.RWMutex{},
-		updateInterval: updateInterval,
-		fetchError:     ErrNoFetchOccurred,
-		client:         client,
+type Option func(*cachingRateService)
+
+func WithClient(client *http.Client) Option {
+	return func(crs *cachingRateService) {
+		crs.client = client
 	}
 }
 
-// Fetches currency data periodically, period is defined by updateInterval.
-// Makes 3 tries before giving up by default
-func (rs *cachingRateService) StartBackgroundTask() {
-	// initial fetch
-	rs.mutex.Lock()
-	rs.currentRate, rs.fetchError = rs.client.GetExchangeRate()
-	rs.mutex.Unlock()
-	go func() {
-		for range time.Tick(rs.updateInterval) {
-			for i := 0; i < RetryCount; i++ {
-				rate, err := rs.client.GetExchangeRate()
-				if err == nil {
-					rs.mutex.Lock()
-					rs.currentRate = rate
-					rs.fetchError = nil
-					rs.mutex.Unlock()
-					break
-				}
-				if i == RetryCount-1 {
-					// Give up
-					rs.mutex.Lock()
-					rs.fetchError = err
-					rs.mutex.Unlock()
-				}
-				time.Sleep(RetryInterval)
-			}
-		}
-	}()
+func WithUpdateInterval(updateInterval time.Duration) Option {
+	return func(crs *cachingRateService) {
+		crs.updateInterval = updateInterval
+	}
 }
 
-func (rs *cachingRateService) GetRate() (float32, error) {
-	rs.mutex.RLock()
-	defer rs.mutex.RUnlock()
-
-	if rs.fetchError != nil {
-		return 0, rs.fetchError
-	}
-	return rs.currentRate, nil
+type NamedFetcher struct {
+	fetcher ExchangeRateFetcher
+	name    string
 }
 
-type httpExchangeRateClient struct {
-	client *http.Client
+func CreateNamedFetcher(fetcher ExchangeRateFetcher, name string) NamedFetcher {
+	return NamedFetcher{fetcher: fetcher, name: name}
 }
```

Comment:
It's seems illogical to me that you should specify fetcher name each time as it remains the same for fetcher of the same time across the project. I suggest to incapsulate name setting in fetcher and just expose `Name()` func that will return it to use outside